openverse_catalog/dags/data_refresh/dag_factory.py

```diff
@@ -0,0 +1,186 @@
+Airflow. Since no two refreshes can run simultaneously, all tasks are run in a
+special `data_refresh` pool with a single worker slot. To ensure that tasks
+run in an acceptable order (ie the trigger step for one DAG cannot run if a
+previously triggered refresh is still running), each DAG has the following
+steps:
+
+1. The `wait_for_data_refresh` step uses a custom Sensor that will wait until
+none of the `external_dag_ids` (corresponding to the other data refresh DAGs)
+are 'running'. A DAG is considered to be 'running' if it is itself in the
+RUNNING state __and__ its own `wait_for_data_refresh` step has completed
+successfully. The Sensor suspends itself and frees up the worker slot if
+another data refresh DAG is running.
+
+2. The `trigger_data_refresh` step then triggers the data refresh by POSTing
+to the `/task` endpoint on the data refresh server with relevant data. A
+successful response will include the `status_check` url used to check on the
+status of the refresh, which is passed on to the next task via XCom.
+
+3. Finally the `wait_for_data_refresh` task waits for the data refresh to be
+complete by polling the `status_url`. Note this task does not need to be
+able to suspend itself and free the worker slot, because we want to lock the
+entire pool on waiting for a particular data refresh to run.
+
+You can find more background information on this process in the following
+issues and related PRs:
+
+- [[Feature] Data refresh orchestration DAG](
+https://github.com/WordPress/openverse-catalog/issues/353)
+"""
+# airflow DAG (necessary for Airflow to find this file)
+import json
+import logging
+from typing import Sequence
+from urllib.parse import urlparse
+
+from airflow import DAG
+from airflow.exceptions import AirflowException
+from airflow.providers.http.operators.http import SimpleHttpOperator
+from airflow.providers.http.sensors.http import HttpSensor
+from common.sensors import ExternalDAGsSensor
+from data_refresh.data_refresh_types import DATA_REFRESH_CONFIGS, DataRefresh
+
+
+logger = logging.getLogger(__name__)
+
+
+XCOM_PULL_TEMPLATE = "{{{{ ti.xcom_pull(task_ids='{}', key='{}') }}}}"
+DATA_REFRESH_POOL = "data_refresh"
+
+
+# Filter for the `trigger_data_refresh` task, used to grab the endpoint needed
+# to poll for the status of the triggered data refresh. This information will
+# then be available via XCom in the downstream tasks.
+def response_filter_data_refresh(response):
+    status_check_url = response.json()["status_check"]
+    return urlparse(status_check_url).path
+
+
+# Response check to the `wait_for_completion` Sensor. Processes the response to
+# determine whether the task can complete.
+def response_check_wait_for_completion(response):
+    data = response.json()
+
+    if data["active"]:
+        # The data refresh is still running. Poll again later.
+        return False
+
+    if data["error"]:
+        raise AirflowException("Error triggering data refresh.")
```

Comment:
Would it make sense to include the contents of `data["error"]` here, or any error info we have? Or is that just True/False?

Reply:
Yeah I'd like to report something more useful here, but we just get back True/False in the response. For the time being I think we can probably rely on the ingestion server's own [error reporting](https://github.com/WordPress/openverse-api/blob/main/ingestion_server/ingestion_server/tasks.py#L130)

Reply:
Ah drat, that sounds good though! You're right, we should be getting sufficient information elsewhere.